### DISCUSSION TITLE
Package rules for `@types/*`

### DISCUSSION THREAD

**Discussion by moltar** [renovatebot/renovate#27531] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Always Latest

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

The `@types/*` packages follow a specific SemVer convention to align the types to the package that is being typed.

It is outlined here: https://github.com/definitelytyped/definitelytyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library

I am wondering if there's a way to configure Renovate to apply these rules for `@types/*` packages.

Effectively, as I understand the `@types/*` major and minor must not exceed the version of the package that is being typed. Can be lower, in the case where types were not updated or have not changed with the package version.

**Comment by rarkins** [comment, marked as answer]
This has been a long-requested topic, although I can't find the original discussion.

It would be particularly complex to achieve because we need to consider the case where _both the package and its types companion are updated at the same time_. So it's not enough to just look at the existing version of `x` before deciding which version of `@types/x` is allowed.

It's also a challenge to implement elegantly in Renovate considering this is very specific to just a subset of packages (`@types/*`) of just one ecosystem (npm).

In short:
- Good idea to do
- We can't so far work out an elegant way to do it

This can be promoted to an issue if/once someone can work out an elegant way to achieve it

**Reply by moltar** [reply]
@rarkins Thanks for the thoughtful response. I do understand that the issue is extremely complex, as I have looked into this too before to even just try to figure it out in a localized tool type of thing, and yeah, it's not so straightforward. I will mark the question as answered. But I am happy to engage in a further discussion with anyone who's interested in figuring this out.
